### PR TITLE
Add an Microprofile OpenAPI example

### PR DESCRIPTION
**Pull request by liweinan**

### CODE REVIEW

**Comment by liweinan** [comment]
@jamezp I'm working on a SmallRye based OpenAPI example, and I've made some progresses. I plan to make the example based on the current `bootable-jar` example(I'll move the openapi example into separate submodules later), because it's convenient to boot the WildFly 
using the `wildfly-jar` plugin IMHO.

Here is the `jaxrs` support provied by the `smallrye-open-api` project:

- https://github.com/smallrye/smallrye-open-api/tree/main/extension-jaxrs

Which must be used in an integrated environment such as Quarkus or WildFly. For example, here is the usage of it in Quakrus:

- https://quarkus.io/guides/openapi-swaggerui

The reason to use it in a container is because the container help to setup the Apis generated by the smallrye framework. For example, in `microprofile-open-api` test code base, here is a setup to enable it in Jdk Http Server:

68E5B583-281B-4AE5-BF45-3D03A411A203

- https://github.com/smallrye/smallrye-open-api/blob/main/testsuite/extra/src/test/java/test/io/smallrye/openapi/tck/ExtraSuiteTestBase.java#L123

And here is the handler in Quarkus:

- https://github.com/quarkusio/quarkus/blob/main/extensions/smallrye-openapi/runtime/src/main/java/io/quarkus/smallrye/openapi/runtime/OpenApiHandler.java#L60

For WildFly, it contains the `openapi` layer as described in:

- https://docs.wildfly.org/21/Galleon_Guide.html#wildfly_galleon_layers

image

So I added the layer in the `pom.xml` of this branch:

```xml
                        <layer>microprofile-openapi</layer>
```

However the plugin reports this error:

```txt
[ERROR] Failed to execute goal org.wildfly.plugins:wildfly-jar-maven-plugin:8.1.0.Final:package (default) on project products-runner: Provisioning failed: Failed to locate layer [model=standalone name=microprofile-openapi] in the the installation feature-pack layout -> [Help 1]
```

I'm not sure how to configure this properly. Could you please provide some help on this? Thanks!

**Comment by jamezp** [comment]
@liweinan I wonder if we should create a resteasy-microprofile-example. The reason this doesn't work is we're using the `wildfly-ee-feature-pack` and we'd need to use the full feature pack that includes the microprofile specs. Talking out loud here, I guess maybe it doesn't make sense to make a new set of examples that is separate. Keeping the examples here probably makes the most sense :)

Anyway, you need change the feature pack to `org.wildfly:wildfly-galleon-pack`.

**Comment by liweinan** [comment]
@jamezp Yes sir! And thanks for the help :D

**Comment by liweinan** [comment]
The setup works:

image

I need to split this example from `bootable-jar-example` into `smallrye-openapi-example`. In addition I need to do some cleanup.

**Comment by liweinan** [comment]
@jamezp I have finished working on the example, could you please help to review it? Thanks

File: pom.xml [hunk 1]
```diff
@@ -40,5 +40,6 @@
         <module>json-binding</module>
         <module>smime</module>
         <module>tracing-example</module>
+        <module>smallrye-openapi</module>
```
Comment (jamezp) [inline pom.xml:43]: This should be moved up two spots to keep these in alphabetical order.

File: smallrye-openapi/pom.xml [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+                            <version>${version.org.jboss.resteasy}</version>
+                        </feature-pack>
+                    </feature-packs>
+                    <layers>
+                        <layer>jaxrs</layer>
+                        <layer>management</layer>
+                        <layer>microprofile-openapi</layer>
+                    </layers>
+                    <excluded-layers>
+                        <layer>deployment-scanner</layer>
```
Comment (jamezp) [inline smallrye-openapi/pom.xml:147]: Is there a specific reason to exclude this?

File: smallrye-openapi/src/main/java/dev/resteasy/examples/openapi/Product.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package dev.resteasy.examples.openapi;
```
Comment (jamezp) [inline smallrye-openapi/src/main/java/dev/resteasy/examples/openapi/Product.java:1]: This should have a copyright header

File: smallrye-openapi/src/main/java/dev/resteasy/examples/openapi/ProductApplication.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package dev.resteasy.examples.openapi;
```
Comment (jamezp) [inline smallrye-openapi/src/main/java/dev/resteasy/examples/openapi/ProductApplication.java:1]: This should have a copyright header

File: smallrye-openapi/src/main/java/dev/resteasy/examples/openapi/ProductResource.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package dev.resteasy.examples.openapi;
```
Comment (jamezp) [inline smallrye-openapi/src/main/java/dev/resteasy/examples/openapi/ProductResource.java:1]: This should have a copyright header

File: smallrye-openapi/src/main/java/dev/resteasy/examples/openapi/ProductApplication.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+import jakarta.ws.rs.core.Application;
+
+import org.eclipse.microprofile.openapi.annotations.OpenAPIDefinition;
+import org.eclipse.microprofile.openapi.annotations.info.Info;
+import org.eclipse.microprofile.openapi.annotations.tags.Tag;
+
+@ApplicationPath("/")
+@OpenAPIDefinition(info = @Info(title = "ProductApplication", version = "1.0.0"), tags = {
+        @Tag(name = "tagX"),
+        @Tag(name = "tagY") })
```
Comment (jamezp) [inline smallrye-openapi/src/main/java/dev/resteasy/examples/openapi/ProductApplication.java:13]: I'm not sure we should have tags here. We're not really using them.

File: smallrye-openapi/src/test/java/dev/resteasy/examples/openapi/OpenApiTestIT.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
+package dev.resteasy.examples.openapi;
```
Comment (jamezp) [inline smallrye-openapi/src/test/java/dev/resteasy/examples/openapi/OpenApiTestIT.java:1]: This should have a copyright header

**Review by jamezp — CHANGES_REQUESTED** [review]
In general we should not name this "smallrye-openapi". SmallRye as an Eclipse MicroProfile OpenAPI implementation. However, here we're using the MicroProfile OpenAPI.

I also question whether or not we should be using the bootable JAR here. If we just use the API's, we could run on WildFly or in other containers. This part doesn't require a change before we can merge this, but it's something we should think about IMO.

**Comment by liweinan** [comment]
@jamezp Thanks for reviewing! I'll modify the PR according to it.

File: smallrye-openapi/pom.xml (hunk 2 above)
Comment (liweinan) [reply smallrye-openapi/pom.xml:147]: it's just copy pasted :) I'll remove it.

**Comment by liweinan** [comment]
@jamezp I reconfigured the example to use `wildfly-plugin` instead of `wildfly-jar-plugin`. After switching the plugin, I can start the wildfly server with:

```bash
$ mvn wildfly:run
...
22:21:17,030 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-5) WFLYSRV0027: Starting deployment of "smallrye-openapi-example.war" (runtime-name: "smallrye-openapi-example.war")
22:21:18,171 INFO  [org.wildfly.extension.microprofile.openapi.smallrye] (MSC service thread 1-8) WFLYMPOAI0004: Registered MicroProfile OpenAPI endpoint '/openapi' for host 'default-host'
22:21:18,392 INFO  [org.jboss.resteasy.resteasy_jaxrs.i18n] (ServerService Thread Pool -- 7) RESTEASY002225: Deploying jakarta.ws.rs.core.Application: class dev.resteasy.examples.openapi.ProductApplication
22:21:18,444 INFO  [org.hibernate.validator.internal.util.Version] (ServerService Thread Pool -- 7) HV000001: Hibernate Validator 8.0.0.Final
22:21:18,473 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 7) WFLYUT0021: Registered web context: '/smallrye-openapi-example' for server 'default-server'
22:21:18,507 INFO  [org.jboss.as.server] (management-handler-thread - 1) WFLYSRV0010: Deployed "smallrye-openapi-example.war" (runtime-name : "smallrye-openapi-example.war")
```

And then I can access the service with:

```bash
➤ curl http://127.0.0.1:8080/openapi                                                                                                                                                                                                                                                                            22:22:00
---
openapi: 3.0.3
info:
  title: ProductApplication
  version: 1.0.0
servers:
- url: /smallrye-openapi-example
paths:
  /products:
    get:
      responses:
        default:
          description: return a list of products
```

However, the `mvn clean install` or `mvn integration-test` failed with:

<img width="1272" alt="image" src="https://user-images.githubusercontent.com/201907/219054318-cd13d4cb-99fc-4dd7-9372-382c34627c82.png">

It seems the project is not deployed:

<img width="2419" alt="image" src="https://user-images.githubusercontent.com/201907/219054757-43bd4001-6752-43b0-832d-47dfb2c2ea62.png">

I guess it's the plugin configuration problem, so the testing doesn't deploy the project. However I don't know how to configure it properly, could you please provide some help on this? Thanks!

**Comment by liweinan** [comment]
After reading this doc:

- http://www.mastertheboss.com/jboss-frameworks/jboss-maven/configuring-maven-wildfly-plugin/ 

I understand how to configure the plugin for testing. The core part is to add this line:

```xml
                            <goal>deploy</goal>
```

**Comment by liweinan** [comment]
@jamezp I have finished modifying the example. Could you please help to review it again? Thanks! (If it's okay I'll do squash and force push)

File: smallrye-openapi/README.adoc [hunk 8]
```diff
@@ -0,0 +1,9 @@
+= RESTEasy SmallRye OpenAPI JAR Example
```
Comment (jamezp) [inline smallrye-openapi/README.adoc:1]: This should be:
```
= RESTEasy OpenAPI Example
```

File: smallrye-openapi/README.adoc [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+
+Demonstrates creating a WildFly bootable JAR with SmallRye OpenAPI enabled.
```
Comment (jamezp) [inline smallrye-openapi/README.adoc:3]: Remove "SmallRye" here as that is an implementation detail.

File: smallrye-openapi/pom.xml [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        <version>2.0.2.Final</version>
+        <relativePath/>
+    </parent>
+
+    <groupId>dev.resteasy.examples</groupId>
+    <artifactId>smallrye-openapi-example</artifactId>
+    <version>6.0.1-SNAPSHOT</version>
+    <packaging>war</packaging>
+    <name>RESTEasy Examples: SmallRye OpenAPI Example</name>
+    <description>RESTEasy SmallRye OpenAPI Example</description>
```
Comment (jamezp) [inline smallrye-openapi/pom.xml:18]: "smallrye" should be removed in all these places.

**Review by jamezp — CHANGES_REQUESTED** [review]
We still need to remove the "smallrye" name as that is an implementation detail. This could just be named "openapi" or "microprofile-openapi", but it shouldn't be named "smallrye-openapi" IMO.

**Comment by liweinan** [comment]
@jamezp Thanks for review! I have finished modifications and squash the commits.

**Review by jamezp — APPROVED** [review]

**Pull request merged**